tools/odbc/result.cpp

```diff
@@ -92,11 +252,336 @@ SQLRETURN SQLGetData(SQLHSTMT statement_handle, SQLUSMALLINT col_or_param_num, S
 			}
 			return SQL_SUCCESS;
 		}
+		case SQL_C_TYPE_DATE: {
+			date_t date;
+			switch (val.type().id()) {
+			case LogicalTypeId::DATE:
+				date = val.GetValue<date_t>();
+				break;
+			case LogicalTypeId::TIMESTAMP:
+			case LogicalTypeId::TIMESTAMP_SEC:
+			case LogicalTypeId::TIMESTAMP_MS:
+			case LogicalTypeId::TIMESTAMP_NS: {
+				try {
+					date = duckdb::CastTimestampToDate::Operation<timestamp_t, date_t>(val.GetValue<timestamp_t>());
```

Comment:
The different timestamps have different units, which the cast does not know. For example, the timestamp `2021-07-15 12:30:00` will be:

```
* 1626352200 in TIMESTAMP_SEC
* 1626352200000 in TIMESTAMP_MS
* 1626352200000000 in TIMESTAMP
* 1626352200000000000 in TIMESTAMP_NS
```

The `CastTimestampToDate` function expects the timestamp in microseconds (i.e. the regular `TIMESTAMP`).
 To correctly support the other scenarios, you can use the following functions:

```
TIMESTAMP_SEC -> Timestamp::FromEpochSeconds
TIMESTAMP_MS  -> Timestamp::FromEpochMs
TIMESTAMP     -> Timestamp::FromEpochMicroSeconds
TIMESTAMP_NS  -> Timestamp::FromEpochNanoSeconds
```

You could e.g. do this by converting the code here into a templated method that takes one of these functions as a template parameter.